### PR TITLE
Break up factories.rb

### PR DESCRIPTION
**Pull request by bencodeorg**
Per some discussion with Nick about setting up a new model [here](https://codedotorg.slack.com/archives/C046G4TRLEN/p1683655188635259), our Rails factories are quite hard to parse, as any model that isn't namespaced is all in a single file (`factories.rb`). Here's a proposal to break these up a bit -- I've broken out a few large chunks into their own files:
- user_factories: 1:1 with a model (this is more of the Rails standard).
- school_factories: a set of models that are are closely affiliated (`School`, `SchoolInfo`, `SchoolDistrict`)
- contact_rollups_factories: ~~a namespace I'm familiar with, should be in its own file per our existing guidance around one file per namespace.~~ Edit: these aren't in their own namespace, just share a common prefix.

End product is that `factories.rb` is about half as big as it was previously.

I think we could also break out factories related to our curriculum models, but I'm less familiar with that area -- hopefully this is a step in the right direction :).

### CODE REVIEW

**Comment by bencodeorg** [comment]
This doesn't seem like an overly controversial change, but tagging @code-dot-org/dev-tools-group in case anyone has thoughts.

**Review by davidsbailey — APPROVED** [review]

**Comment by Hamms** [comment]
I'm a little conflicted here. On the one hand, having such a mammoth factories file is certainly unwieldy. On the other hand, I don't love how loosely defined the new categories system is. I feel like now when I'm looking at a particular model and want to go look at the factory for the model, I'm going to have to go look at all the different factory files to figure out if any of the chunks we've decided to break off are likely to contain the relevant factory or if it's still in the default file.

What do we think about leaning even *harder* into the refactor, and switching entirely over to the more-Rails-aligned standard of having a `test/factories/model.rb` factory file for each `app/model.rb` model file?

**Comment by bencodeorg** [comment]
> I'm a little conflicted here. On the one hand, having such a mammoth factories file is certainly unwieldy. On the other hand, I don't love how loosely defined the new categories system is. I feel like now when I'm looking at a particular model and want to go look at the factory for the model, I'm going to have to go look at all the different factory files to figure out if any of the chunks we've decided to break off are likely to contain the relevant factory or if it's still in the default file.
> 
> What do we think about leaning even _harder_ into the refactor, and switching entirely over to the more-Rails-aligned standard of having a `test/factories/model.rb` factory file for each `app/model.rb` model file?

Yeah that's fair. I feel like that's probably the right longer term direction (splitting everything up). FWIW, the new files I think lean towards the "right" direction:
- user: we already have a workshop_factories file that aligns 1:1 with a model (albeit one that is in a namespace). This is the biggest chunk (and I think would be clear enough in its own file), and would move towards that full split you describe.
- contact_rollups and school: these are a funky new pattern, as they encompasses multiple models that aren't in the same namespace. I could move these back to factories.rb if you'd prefer -- all of the models in a given file at least all start with the prefix, so I think it's reasonably clear? Or I could break them up further to kick off that broader refactor?

**Comment by Hamms** [comment]
Definitely agreed that this is a step in the right direction, I just don't love how much it leaves us in an in-between state. I would be happier with it if the contact_rollups and school factories were broken up into one model per factory file, though! But what would make me really happy is for all the rest of the megafactory to be similarly dispersed.

**Comment by nicklathe** [comment]
I think I agree with @Hamms, either a single file (like it is today) or a separate model per factory file (the rails auto gen way) feels like it helps with discoverability when someone is trying to find a specific factory. However if this is a step towards breaking things up further, I'm onboard.

**Comment by nicklathe** [comment]
Thinking through the LTI lens (as that's I'm working on), we are adding three new models: lti_integrations, lti_users, and lti_deployments. These all feel like they could live in the school_factories.rb (since they are related to a school/district LMS), with the exception of lti_users. Based on naming alone, it feels like it would go in the user_factories.rb, but now it lives in a separate file as the other LTI related factories. Thoughts?

**Comment by davidsbailey** [comment]
> Thinking through the LTI lens (as that's I'm working on), we are adding three new models: lti_integrations, lti_users, and lti_deployments. These all feel like they could live in the school_factories.rb (since they are related to a school/district LMS), with the exception of lti_users. Based on naming alone, it feels like it would go in the user_factories.rb, but now it lives in a separate file as the other LTI related factories. Thoughts?

Nick, this question supports Elijah's point that in this in-between state, it's hard to know where to put new factories. if we are committing to the direction of separate factories, then shouldn't the new factories go in per-model files, like the ones created by rails by default when you generate the model?

Given the concerns about the in-between state, this seems like a good case in point on how to direct developers on where new factories should go.
